src/server/engine_shard.cc

```diff
@@ -620,7 +700,43 @@ void EngineShard::Heartbeat() {
     }
     return;
   }
-  start = std::chrono::system_clock::now();
+
+  thread_local bool check_huffman = (shard_id_ == 0);  // run it only on shard 0.
+  if (check_huffman) {
+    auto* ptr = db_slice.GetDBTable(0);
+    if (ptr) {
+      size_t key_usage = ptr->stats.memory_usage_by_type[OBJ_KEY];
+      size_t obj_usage = ptr->stats.obj_memory_usage;
+
+#ifdef NDEBUG
+#define MB_THRESHOLD (50 * 1024 * 1024)
+#else
+#define MB_THRESHOLD (5 * 1024 * 1024)
+#endif
+
+      if (key_usage > MB_THRESHOLD && key_usage > obj_usage / 8) {
+        VLOG(1) << "Scheduling huffman check task, key usage: " << key_usage
+                << ", obj usage: " << obj_usage;
+
+        check_huffman = false;  // trigger only once.
+
+        // launch the task
+        HuffmanCheckTask* task = new HuffmanCheckTask();
+        huffman_check_task_id_ = ProactorBase::me()->AddOnIdleTask([task] {
+          if (!shard_ || !namespaces) {
+            delete task;
+            return -1;
+          }
```

Comment:
It feels a little sloppy overall, like it was written for a POC. If it becomes more complicated, it would be preferrably to refactor it a little. For example, you don't need the pointer, it can be just a member of the lambda,